src/infill/SierpinskiFill.cpp

```diff
@@ -0,0 +1,802 @@
+{
+    float ret = 0;
+    for (auto it = begin; it != end; ++it)
+    {
+        SierpinskiTriangle* node = *it;
+        ret += node->getSubdivisionError();
+    }
+    return ret;
+}
+
+
+
+
+void SierpinskiFill::debugOutput(SVG& svg)
+{
+    svg.writePolygon(aabb.toPolygon(), SVG::Color::RED);
+
+    // draw triangles
+    for (SierpinskiTriangle* node : sequence)
+    {
+        SierpinskiTriangle& triangle = *node;
+        svg.writeLine(triangle.a, triangle.b, SVG::Color::GRAY);
+        svg.writeLine(triangle.a, triangle.straight_corner, SVG::Color::GRAY);
+        svg.writeLine(triangle.b, triangle.straight_corner, SVG::Color::GRAY);
+    }
+}
+
+
+SierpinskiFill::Edge SierpinskiFill::SierpinskiTriangle::getFromEdge()
+{
+    Edge ret;
+    switch(dir)
+    {
+        case SierpinskiDirection::AB_TO_BC:
+            ret = Edge{a, b};
+            break;
+        case SierpinskiDirection::AC_TO_AB:
+            ret = Edge{straight_corner, a};
+            break;
+        case SierpinskiDirection::AC_TO_BC:
+            ret = Edge{straight_corner, a};
+            break;
+    }
+    if (!straight_corner_is_left)
+    {
+        std::swap(ret.l, ret.r);
+    }
+    return ret;
+}
+
+SierpinskiFill::Edge SierpinskiFill::SierpinskiTriangle::getToEdge()
+{
+    Edge ret;
+    switch(dir)
+    {
+        case SierpinskiDirection::AB_TO_BC:
+            ret = Edge{straight_corner, b};
+            break;
+        case SierpinskiDirection::AC_TO_AB:
+            ret = Edge{b, a};
+            break;
+        case SierpinskiDirection::AC_TO_BC:
+            ret = Edge{straight_corner, b};
+            break;
+    }
+    if (!straight_corner_is_left)
+    {
+        std::swap(ret.l, ret.r);
+    }
+    return ret;
+}
+
+
+Polygon SierpinskiFill::generateCross() const
+{
+    Polygon ret;
+
+    for (SierpinskiTriangle* max_level_it : sequence)
+    {
+        SierpinskiTriangle& triangle = *max_level_it; 
+        Point edge_middle = triangle.a + triangle.b + triangle.straight_corner;
+        switch(triangle.dir)
+        {
+            case SierpinskiTriangle::SierpinskiDirection::AB_TO_BC:
+                edge_middle -= triangle.a;
+                break;
+            case SierpinskiTriangle::SierpinskiDirection::AC_TO_AB:
+                edge_middle -= triangle.straight_corner;
+                break;
+            case SierpinskiTriangle::SierpinskiDirection::AC_TO_BC:
+                edge_middle -= triangle.a;
+                break;
+        }
+        ret.add(edge_middle / 2);
+    }
+
+    float realized_length = INT2MM(ret.polygonLength());
+    float requested_length = root.requested_length;
+    float error = (realized_length - requested_length) / requested_length;
+    std::cerr << "realized_length: " << realized_length << ", requested_length: " << requested_length << "  :: " << (.01*(int(10000*error))) << "% error\n";
```

Comment:
Logging!

Reply:
yeah yeah. I've done that ages ago! Like 10 comments up already or something! :P